### ISSUE TITLE
Simplify diff when moving an item to a different inherent impl while changing it

### ISSUE THREAD

**Issue by joshka** [cargo-public-api/cargo-public-api#467]
For ratatui, I'm seeing a lot of duplicate lines in the output between 0.21.0 and 0.22.0:

```diff
❯ cargo public-api diff v0.21.0..v0.22.0

Removed items from the public API
=================================
-pub fn ratatui::layout::Layout::direction(self, direction: ratatui::layout::Direction) -> ratatui::layout::Layout
-pub fn ratatui::layout::Layout::horizontal_margin(self, horizontal: u16) -> ratatui::layout::Layout
-pub fn ratatui::layout::Layout::margin(self, margin: u16) -> ratatui::layout::Layout
-pub fn ratatui::layout::Layout::vertical_margin(self, vertical: u16) -> ratatui::layout::Layout

... much later in the added items section

+pub const fn ratatui::layout::Layout::direction(self, direction: ratatui::layout::Direction) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::direction(self, direction: ratatui::layout::Direction) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::direction(self, direction: ratatui::layout::Direction) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::horizontal_margin(self, horizontal: u16) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::horizontal_margin(self, horizontal: u16) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::horizontal_margin(self, horizontal: u16) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::margin(self, margin: u16) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::margin(self, margin: u16) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::margin(self, margin: u16) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::new() -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::new() -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::new() -> ratatui::layout::Layout
+pub fn ratatui::layout::Layout::split(&self, area: ratatui::layout::Rect) -> alloc::rc::Rc<[ratatui::layout::Rect]>
+pub fn ratatui::layout::Layout::split(&self, area: ratatui::layout::Rect) -> alloc::rc::Rc<[ratatui::layout::Rect]>
+pub const fn ratatui::layout::Layout::vertical_margin(self, vertical: u16) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::vertical_margin(self, vertical: u16) -> ratatui::layout::Layout
+pub const fn ratatui::layout::Layout::vertical_margin(self, vertical: u16) -> ratatui::layout::Layout
```
(full output at https://github.com/tui-rs-revival/ratatui/wiki/Migrating-from-TUI)

Adding the `--simplified` option (or `-sss`) does not change this.
I'd also expect these to show up in the changed items list, not as a remove and add.

**Comment by Emilgardis** [comment]
A change from non-const to const should indeed probably be annotated under changed items

**Comment by joshka** [comment]
Yeah, there's a few big parts of that release - added a prelude, so I wonder if the extra pub re-exports are doing something weird?

**Comment by Enselic** [comment]
Thank you for reporting. Not sure what is going on here. I did a quick test, and diffing between

```rust
pub fn a() {}
pub fn b() {}
pub fn c() {}
```

```rust
pub fn a() {}
pub const fn b() {}
pub fn c() {}
```

gives the expected output:

```
$ cargo public-api diff HEAD^..HEAD
Removed items from the public API
=================================
(none)

Changed items in the public API
===============================
-pub fn lib::b()
+pub const fn lib::b()

Added items to the public API
=============================
(none)
```

I don't think re-exports with a prelude should matter here. When diffing, the rendered items are used, so the differ does not see a difference between re-exports and original items. It just sees two items with different paths.

Would be good if we could find a minimal reproducer for this bug. With a MCVE it is probably easy to fix.

**Comment by joshka** [comment]
A more narrow example (not of the const part, but the duplication part):
For commit https://github.com/tui-rs-revival/ratatui/commit/358b50b
```
❯ cargo public-api -sss diff 358b50b~..358b50b
...
Changed items in the public API
===============================
-pub struct ratatui::style::Modifier
+pub struct ratatui::style::Modifier(_)
-pub struct ratatui::widgets::Borders
+pub struct ratatui::widgets::Borders(_)

Added items to the public API
=============================
+impl ratatui::style::Modifier
+impl ratatui::style::Modifier
...
+impl ratatui::widgets::Borders
+impl ratatui::widgets::Borders
```
Guessing those were both added by macros and there probable is two impl sections. But the public api should effectively treat that as one section. I've not dug into the cargo-public api code yet, but if multiple impl sections are being treated as different parts of the API, that would probably lead to some of this oddness.

**Comment by joshka** [comment]
MCVE:
```
❯ git diff HEAD~
diff --git a/src/lib.rs b/src/lib.rs
index ce34135..99d7a35 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -4,5 +4,8 @@ impl Foo {
     pub fn new() -> Foo {
         Foo
     }
-    pub fn bar(&self) {}
+}
+
+impl Foo {
+    pub const fn bar(&self) {}
 }

❯ cargo public-api -sss diff HEAD~..HEAD
 Documenting spike v0.1.0 (/Users/joshka/local/spike)
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
 Documenting spike v0.1.0 (/Users/joshka/local/spike)
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
Removed items from the public API
=================================
-pub fn spike::Foo::bar(&self)

Changed items in the public API
===============================
(none)

Added items to the public API
=============================
+impl spike::Foo
+pub const fn spike::Foo::bar(&self)
```

**Comment by Enselic** [comment]
Thank you for the help with the MCVE. Turns out this only has to do with moving an item between inherent impls, and not with `const`ness. I can reproduce the problem by diffing between this public API:

```rust
pub struct S;

impl S {
    pub fn first_impl() {}

    pub fn f() {}
}

impl S {
    pub fn second_impl() {}
}
```

and this public API:

```rust
pub struct S;

impl S {
    pub fn first_impl() {}
}

impl S {
    pub fn second_impl() {}

    pub fn f(_arg: bool) {}
}
```

which results in this diff:

```console
$ cargo public-api diff HEAD~..HEAD
Removed items from the public API
=================================
-pub fn lib::S::f()

Changed items in the public API
===============================
(none)

Added items to the public API
=============================
+pub fn lib::S::f(_arg: bool)
```

I think I'll change this to an enhancement, because the output is not wrong/inaccurate. I'm afraid it won't be as easy as I thought initially to fix this, because `impl`s are by their nature quite special, and we have some hacks here and there to make them appear correctly in the output. I'm afraid that fixing this will cause regressions in other places. (And if so, these regressions will not go undetected, because we have very solid and comprehensive regression tests in this project).

**Comment by joshka** [comment]
I took a brief look through the code and understand the implementation of this might not be simple. A question mainly for my curiosity. Is the public api of a rust type with two `impl` sections both those sections, or just one? I understand that the docs are rendered with multiple sections, but is this actually something that the type exposes, or is it just an implementation detail that's useful for arranging code and docs? I'd assumed the latter, but your comment seems to imply the former.

**Comment by Emilgardis** [comment]
The problem is when the impl contains generics/where clauses, and where I suspect this goes wrong, we're being overly pessimistic about the "sameness" of an impl.

As for public api terms, moving a function from one impl to another identical impl doesn't change the public api at all, if the `impl` is exactly the same it's only an implementation detail (pun intended). I disagree with the notion that this is not a bug/wrong output.

**Comment by Emilgardis** [comment]
Ideally, I think the output for https://github.com/Enselic/cargo-public-api/issues/467#issuecomment-1644569649 should be

```
$ cargo public-api diff HEAD~..HEAD
Removed items from the public API
=================================
(none)

Changed items in the public API
===============================
impl lib::S
-pub fn lib::S::f()
+pub fn lib::S::f(_arg: bool)


Added items to the public API
=============================
(none)
```

(note the presence of `impl lib::S`)

**Comment by Enselic** [comment]
I agree that ideally `f()` should be under "changed items", but I'm not so sure about having an item under "changed items" that has not been changed (i.e. `impl lib::S` which is an item by itself).

If we change this back to bug, can we maybe label it low-prio in that case? Motivation would be:
* The output is not wrong, just unnecessarily verbose
* It's a bit of an edge case to move items between impls and changing them at the same time
* To someone glancing over our open issues, I want to make it clear on first glance that we have no known bugs that are not low-prio. I don't like unfixed bugs, but I don't think I'll find the time and motivation to fix this one (but I would be happy to mentor and review code)

>  I understand that the docs are rendered with multiple sections, but is this actually something that the type exposes, or is it just an implementation detail that's useful for arranging code and docs?

I remember going back and forth between combining multiple inherent impls into one when rendering. IIRC we ended up in the current situation because
* that's what rustdoc HTML does, and we try to look like rustdoc HTML
* typically multiple inherent impls have different generics arguments, which needs to be rendered separately anyway

**Comment by joshka** [comment]
As a user of this tool, I don't mind whether this labeled as a bug or a feature. The distinction only matters for those that see things from the perspective of judgment rather than from appreciation. This is an awesome tool and it would be even better if multiple impls were seen as the same rather than different when detecting changes.

Looking at the implementation the other day, I noticed that this is probably not a simple change to implement because the processing of the diffs happens as the impls are encountered rather than getting all the possible impls and then processing them afterwards. (I only briefly looked at this, so perhaps I'm mistaken on that assessment). My concern over fixing this was that there didn't seem like an easy neat solution without rewriting some of that logic.

**Comment by Enselic** [comment]
Let's go ahead and close this as won't fix, because I don't think anyone will ever bother to fix this, and I like a clean issue inbox.
